### PR TITLE
[pylint] - add `unnecessary-list-index-lookup` (`PLR1736`) + autofix

### PR DESCRIPTION
**Pull request by diceroll123**
## Summary

Add [R1736](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/unnecessary-list-index-lookup.html) along with the autofix

See #970 

## Test Plan

`cargo test` and manually

### CODE REVIEW

**Comment by Skylion007** [comment]
@diceroll123 I tried out your rule and found a bug:

Consider this code snippit
```python
a = list(range(5, 10))
for i, j in enumerate(a):
    a[i] = j + 1
    assert a[i] > j
```
becomes transformed to:
```python
a = list(range(5, 10))
for i, j in enumerate(a):
    a[i] = j + 1
    assert j > j
```
which becomes now outputs an Assertion Error. In other words, you have to make sure `a[i]` is always an alias of `j` which is not the case here. I suspect your other PR has a similar issue.

**Comment by diceroll123** [comment]
@Skylion007 good catch, thanks! Will update.

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        return None;
+    };
+
+    // Check that the function is the `enumerate` builtin.
+    let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
+        return None;
+    };
+    if id != "enumerate" {
+        return None;
+    };
```
Comment (zanieb) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:222]: Should this use `checker.semantic().resolve_call_path(func)`?

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+use ruff_python_ast::visitor::Visitor;
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for uses of enumeration and accessing the value by index lookup.
+///
+/// ## Why is this bad?
+/// The value is already accessible by the 2nd variable from the enumeration.
```
Comment (zanieb) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:15]: Perhaps
```suggestion
/// It is more succinct to use the variable for the value at the current index which is already in scope from the iterator. 
```

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::visitor;
+use ruff_python_ast::visitor::Visitor;
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for uses of enumeration and accessing the value by index lookup.
```
Comment (zanieb) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:12]: Perhaps
```suggestion
/// Checks for access of a list item at the current index when using enumeration.
```

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+/// for index, letter in enumerate(letters):
+///     print(letter)
+/// ```
+#[violation]
+pub struct UnnecessaryListIndexLookup;
+
+impl AlwaysFixableViolation for UnnecessaryListIndexLookup {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Unnecessary list index lookup")
```
Comment (zanieb) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:38]:
```suggestion
        format!("Unnecessary lookup of list item by index")
```

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    fn fix_title(&self) -> String {
+        format!("Remove unnecessary list index lookup")
```
Comment (zanieb) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:42]:
```suggestion
        format!("Use existing item variable instead")
```

File: crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        print(letters[index])  # PLR1736
+        blah = letters[index]  # PLR1736
+        assert letters[index]  == "d"  # PLR1736
+
+
+def dont_fix_these():
+    # once there is an assignment to the sequence[index], we stop emitting diagnostics
+    for index, letter in enumerate(letters):
+        letters[index] = "d"  # Ok
+        assert letters[index] == "d"  # Ok
```
Comment (zanieb) [inline crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py:19]: What about a case where `index` is mutated? e.g. `index += 1`

**Review by zanieb — COMMENTED** [review]
I think @charliermarsh  is better suited to review the implementation, but I read through everything as best as I could :)

The ecosystem changes look good.

File: crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py (hunk 6 above)
Comment (diceroll123) [reply crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py:19]: Adding a case for that, as well as the sequence itself being mutated! Thanks!

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs (hunk 1 above)
Comment (diceroll123) [reply crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs]: Good change, I've also added detection for `builtins.enumerate` along with it, thanks!

File: crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import builtins
```
Comment (dhruvmanila) [inline crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_list_index_lookup.py:1]: These tests are really thorough! Thanks for thinking through them.

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        })
+        | Expr::SetComp(ast::ExprSetComp {
+            elt, generators, ..
+        })
+        | Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            for comp in generators {
+                if let Some((sequence, index_name, value_name)) =
+                    enumerate_items(checker, &comp.iter, &comp.target)
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:202]: nit: using `let Some(..) = .. else { continue; }` will avoid having an additional indent

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            }) => {
+                if let Expr::Name(ast::ExprName { id, .. }) = value.as_ref() {
+                    if id == self.sequence_name {
+                        if let Expr::Name(ast::ExprName { id, .. }) = slice.as_ref() {
+                            if id == self.index_name {
+                                self.diagnostic_ranges.push(*range);
+                            }
+                        }
+                    }
+                }
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:104]: nit: can we avoid having nested `if` conditions by using `let Some(ast::ExprName { id, .. }) = value.as_name_expr() else { return; }`

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+                    Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                        if let Expr::Name(ast::ExprName { id, .. }) = value.as_ref() {
+                            if id == self.sequence_name {
+                                if let Expr::Name(ast::ExprName { id, .. }) = slice.as_ref() {
+                                    if id == self.index_name {
+                                        return true;
+                                    }
+                                }
+                            }
+                        }
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:147]: Same as above

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    checker: &mut Checker,
+    call_expr: &Expr,
+    tuple_expr: &Expr,
+) -> Option<(String, String, String)> {
+    let Expr::Call(ast::ExprCall {
+        func, arguments, ..
+    }) = call_expr
+    else {
+        return None;
+    };
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:235]: We could use the `call_expr.as_call_expr()?` pattern throughout this function. For example, the highlighted code would become:

```rust
let Some(ast::ExprCall { func, arguments, .. }) = call_expr.as_call_expr()?;
```

**Review by dhruvmanila — APPROVED** [review]
Overall this is pretty good. Thanks for the thorough test cases. I just have a few nits but those shouldn't be a blocker to merge this PR.

**Comment by diceroll123** [comment]
Will address those, all good feedback, thanks @dhruvmanila!

File: crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs (hunk 11 above)
Comment (diceroll123) [reply crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs:235]: Looks like the right code was

```rust
let ast::ExprCall {
    func, arguments, ..
} = call_expr.as_call_expr()?;
```

That tripped me up for a sec! 🤣 Thanks for the push in the right direction though!

Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs]: Ah right, sorry!

**Pull request merged**